File: rust/src/benchmark/prepare_operations.rs
use crate::config::Config;
use crate::model::domain::{Cost, Customer, Material, Operation, Period, Purchase, Sale, Vendor};
use rand::random_range;
use sqlx::types::Decimal;

pub fn prepare_operations(period: Period, config: &Config) -> Vec<Operation> {
    let mut operations = Vec::with_capacity(config.operations);
    
    for _ in 0..config.operations * 20 / 100 {
        let order = random_range(0..config.operations);
        let operation = Operation::PeriodReport(period);
        operations.push((operation, order));
    }

    for _ in 0..config.operations * 10 / 100 {
        let order = random_range(0..config.operations);
        let operation =  Operation::YearReport(period);
        operations.push((operation, order));
    } 
    for n in 0..config.operations * 20 / 100 {
        let order = random_range(0..config.operations);
        let posting_date = period.date(order as u32 * period.last_day() / config.operations as u32);
        let operation =  Operation::Cost(Cost {
            material: Material(random_range(0..config.materials)),
            vendor: Vendor(random_range(0..config.vendors)),
            amount: Decimal::from(random_range(1000..2000)),
            doc_no: format!("COST-{}-{:08}", period, n),
            posting_date,
        });
        operations.push((operation, order));
    }

    for n in 0..config.operations * 25 / 100 {
        let order = random_range(0..config.operations);
        let posting_date = period.date(order as u32 * period.last_day() / config.operations as u32);
        let operation =  Operation::Purchase(Purchase {
            material: Material(random_range(0..config.materials)),
            vendor: Vendor(random_range(0..config.vendors)),
            quantity: Decimal::from(random_range(1000..2000)),
            price: Decimal::from(random_range(100..200)),
            doc_no: format!("PURCH-{}-{:08}", period, n),
            posting_date,
        });
        operations.push((operation, order));
    } 
    
    for n in 0..config.operations * 25 / 100 {
        let order = random_range(0..config.operations);
        let posting_date = period.date(order as u32 * period.last_day() / config.operations as u32);
        let operation =  Operation::Sale(Sale {
            material: Material(random_range(0..config.materials)),
            customer: Customer(random_range(0..config.customers)),
            quantity: Decimal::from(random_range(100..200)),
            sale_doc_no: format!("SALE-{}-{:08}", period, n),
            cogs_doc_no: format!("COGS-{}-{:08}", period, n),
            posting_date,
        });
        operations.push((operation, order));
    }
    
    operations.push((Operation::OpenPeriod(period), config.operations));
    let close_order = random_range(1..=5) * config.operations / period.last_day() as  usize;
    operations.push((Operation::ClosePeriod(period), close_order));
    operations.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
    let operations = operations
        .into_iter()
        .map(|(operation, _)| operation)
        .collect::<Vec<Operation>>();
    operations
}

File: golang/benchmark/prepare_operations.go
package benchmark

import (
	config2 "bench-pg-go/config"
	"bench-pg-go/model/domain"
	"cmp"
	"fmt"
	"github.com/shopspring/decimal"
	"slices"
)

type PreOperation struct {
	operation domain.Operation
	order     int
}

func prepareOperations(period domain.Period, config *config2.Config) []domain.Operation {
	ops := make([]PreOperation, 0, config.Operations)

	for i := 0; i < config.Operations*20/100; i++ {
		order := randGen.Intn(config.Operations)
		operation := domain.NewPeriodReport(period)
		ops = append(ops, PreOperation{operation, order})
	}

	for i := 0; i < config.Operations*10/100; i++ {
		order := randGen.Intn(config.Operations)
		operation := domain.NewYearReport(period)
		ops = append(ops, PreOperation{operation, order})
	}

	for i := 0; i < config.Operations*20/100; i++ {
		order := randGen.Intn(config.Operations)
		postingDate := period.Date(order * period.LastDay() / config.Operations)
		operation := domain.NewCost(
			domain.NewMaterial(randGen.Intn(config.Materials)),
			domain.NewVendor(randGen.Intn(config.Vendors)),
			decimal.NewFromInt(int64(randGen.Intn(1000)+1000)),
			fmt.Sprintf("COST-%s-%08d", period.String(), i),
			postingDate,
		)
		ops = append(ops, PreOperation{operation, order})
	}

	for i := 0; i < config.Operations*25/100; i++ {
		order := randGen.Intn(config.Operations)
		postingDate := period.Date(order * period.LastDay() / config.Operations)
		operation := domain.NewPurchase(
			domain.NewMaterial(randGen.Intn(config.Materials)),
			domain.NewVendor(randGen.Intn(config.Vendors)),
			decimal.NewFromInt(int64(randGen.Intn(1000)+1000)),
			decimal.NewFromInt(int64(randGen.Intn(100)+100)),
			fmt.Sprintf("PURCH-%s-%08d", period.String(), i),
			postingDate,
		)
		ops = append(ops, PreOperation{operation, order})
	}

	for i := 0; i < config.Operations*25/100; i++ {
		order := randGen.Intn(config.Operations)
		postingDate := period.Date(order * period.LastDay() / config.Operations)
		operation := domain.NewSale(
			domain.NewMaterial(randGen.Intn(config.Materials)),
			domain.NewCustomer(randGen.Intn(config.Customers)),
			decimal.NewFromInt(int64(randGen.Intn(100)+100)),
			fmt.Sprintf("SALE-%s-%08d", period.String(), i),
			fmt.Sprintf("COGS-%s-%08d", period.String(), i),
			postingDate,
		)
		ops = append(ops, PreOperation{operation, order})
	}

	ops = append(ops, PreOperation{domain.NewOpenPeriod(period), config.Operations})
	closeOrder := (randGen.Intn(4) + 1) * config.Operations / period.LastDay()
	ops = append(ops, PreOperation{domain.NewClosePeriod(period), closeOrder})

	slices.SortFunc(ops, func(a, b PreOperation) int {
		return cmp.Compare(a.order, b.order)
	})

	operations := make([]domain.Operation, 0, len(ops))
	for _, op := range ops {
		operations = append(operations, op.operation)
	}
	return operations
}
